### PR TITLE
Check that procedures of a generic are distinguishable

### PR DESCRIPTION
**Pull request by tskeith**
Specific procedures of a generic identifier must be pair-wise
distinguishable so that a call to the generic can be resolved to at
most one of the specific procedures. Constraint C1514 specifies the
rules that distinguish them.

This change implements that constraint, except for the parts related
to passed-object dummy data arguments.

Notes:
- Almost all of the logic is currently in `DistinguishUtils` in
  `characteristics.cc`. We will probably need to share some of it for
  resolving generics, so it may end up somewhere else.
- This implementation tries to follow the standard as closely as
  possible so is not the most efficient. We may want to revisit that
  decision too.

### CODE REVIEW

File: lib/evaluate/type.cc [hunk 1]
```diff
@@ -163,10 +199,8 @@ std::optional<DynamicType> DynamicType::From(
   } else if (const auto *derived{type.AsDerived()}) {
     return DynamicType{
         *derived, type.category() == semantics::DeclTypeSpec::ClassDerived};
-  } else if (type.category() == semantics::DeclTypeSpec::ClassStar) {
-    return DynamicType::UnlimitedPolymorphic();
   } else {
-    // Assumed-type dummy arguments (TYPE(*)) do not have dynamic types.
+    return DynamicType::UnlimitedPolymorphic();
```
Comment (tskeith) [inline lib/evaluate/type.cc]: We need to have a dynamic type for `TYPE(*)` when determining whether two procedures are distinguishable.

Comment (klausler) [reply lib/evaluate/type.cc]: On the way!  Will get it into the PR I'm about to merge.

File: lib/evaluate/characteristics.cc [hunk 2]
```diff
@@ -216,210 +426,247 @@ std::ostream &Procedure::Dump(std::ostream &o) const {
...
+    const DummyProcedure &x, const DummyProcedure &y) {
+  const Procedure &xProc{x.procedure.value()};
+  const Procedure &yProc{y.procedure.value()};
+  if (Distinguishable(xProc, yProc)) {
+    return true;
+  } else {
+    const std::optional<FunctionResult> &xResult{xProc.functionResult};
+    const std::optional<FunctionResult> &yResult{yProc.functionResult};
+    return xResult ? !yResult || Distinguishable(*xResult, *yResult)
+                   : yResult.has_value();
```
[the remark this answers is not in the document]
Comment (klausler) [reply lib/evaluate/characteristics.cc]: A dummy argument can be `EXTERNAL` and never be called, just passed along as an actual argument.  It is known to be a procedure, but nothing else is known about its interface.

File: lib/evaluate/intrinsics.cc [hunk 3]
```diff
@@ -1392,7 +1391,7 @@ SpecificCall IntrinsicProcTable::Implementation::HandleNull(
               typeAndShape.emplace(*type);
             }
             characteristics::DummyDataObject ddo{typeAndShape.value()};
-            args.emplace_back(std::move(ddo));
+            args.emplace_back("mold", std::move(ddo));
```
[the remark this answers is not in the document]
Comment (klausler) [reply lib/evaluate/intrinsics.cc]: Does it matter?  A `const char *` is converted to a string.  I don't have a problem with that.

File: lib/evaluate/type.cc [hunk 4]
```diff
@@ -146,6 +146,42 @@ bool DynamicType::IsTypeCompatibleWith(const DynamicType &that) const {
...
+  }
+  return true;
+}
+
+bool DynamicType::IsTkCompatibleWith(const DynamicType &that) const {
+  if (category_ != that.category_ || kind_ != that.kind_) {
+    return false;
+  } else if (category_ != TypeCategory::Derived) {
+    return true;  // same intrinsic category and kind
+  } else if (IsUnlimitedPolymorphic()) {
```
[the remark this answers is not in the document]
Comment (klausler) [reply lib/evaluate/type.cc]: `TYPE(*)` is unlimited polymorphic in the standard, so let's hew to that definition.  We can distinguish assumed-type from `CLASS(*)` if we need to do so.

File: lib/evaluate/characteristics.cc [hunk 5]
```diff
@@ -189,24 +284,139 @@ bool FunctionResult::operator==(const FunctionResult &that) const {
...
+                  return std::nullopt;
+                }
+                auto resultType{DynamicType::From(*type)};
+                if (!resultType) {
+                  return std::nullopt;
+                }
+                result.functionResult = FunctionResult{*resultType};
+              } else {
+                // subroutine, not function
+                if (interface.type() != nullptr) {
```
[the remark this answers is not in the document]
Comment (tskeith) [reply lib/evaluate/characteristics.cc]: I'm not sure -- I need to look for an example that goes through this case. This isn't part of my change, it was merged in.

File: lib/evaluate/type.cc (hunk 4 above)
[the remark this answers is not in the document]
Comment (tskeith) [reply lib/evaluate/type.cc]: `TYPE(*)` is unlimited polymorphic. There is an example at the end of `resolve53.f90`.

File: lib/evaluate/type.cc [hunk 6]
```diff
@@ -146,6 +146,42 @@ bool DynamicType::IsTypeCompatibleWith(const DynamicType &that) const {
       (isPolymorphic_ && IsAncestorTypeOf(derived_, that.derived_));
 }
 
+// Do the kind type parameters of type1 have the same values as the
+// corresponding kind type parameters of the type2?
+static bool IsKindCompatible(const semantics::DerivedTypeSpec &type1,
```
[the remark this answers is not in the document]
Comment (tskeith) [reply lib/evaluate/type.cc]: Yes, I'll look into doing that.

File: lib/evaluate/intrinsics.cc (hunk 3 above)
[the remark this answers is not in the document]
Comment (tskeith) [reply lib/evaluate/intrinsics.cc]: Will do

File: lib/evaluate/characteristics.cc [hunk 7]
```diff
@@ -216,210 +426,247 @@ std::ostream &Procedure::Dump(std::ostream &o) const {
...
+  });
 }
 
-Procedure::Procedure(FunctionResult &&fr, DummyArguments &&args, Attrs a)
-  : functionResult{std::move(fr)}, dummyArguments{std::move(args)}, attrs{a} {}
-Procedure::Procedure(DummyArguments &&args, Attrs a)
-  : dummyArguments{std::move(args)}, attrs{a} {}
+// Return the number of dummy data objects in args that are not
+// distinguishable from x and not passed-object.
+int DistinguishUtils::CountDistinguishableFrom(
```
[the remark this answers is not in the document]
Comment (tskeith) [reply lib/evaluate/characteristics.cc]: Yes, much better.

File: lib/evaluate/intrinsics.cc (hunk 3 above)
[the remark this answers is not in the document]
Comment (tskeith) [reply lib/evaluate/intrinsics.cc]: > Does it matter? A `const char *` is converted to a string. I don't have a problem with that.

I added `"mold"s` in the other place. They should be the same, but I agree it doesn't matter which way.

File: lib/evaluate/characteristics.cc [hunk 8]
```diff
@@ -88,6 +84,16 @@ std::optional<TypeAndShape> TypeAndShape::Characterize(
   }
 }
 
+bool TypeAndShape::IsCompatibleWith(
+    parser::ContextualMessages &messages, const TypeAndShape &that) const {
+  if (!type_.IsTypeCompatibleWith(that.type_)) {
```
Comment (klausler) [inline lib/evaluate/characteristics.cc]: Consider inverting this test and turning the body into an if/else structure, it might be more readable.

**Comment by tskeith** [comment]
> I had to Google "TKR", and all of the hits I got were for total knee replacements. Googling "TKR Fortran" yielded more relevant results.

It's defined in the standard in 15.4.3.4.5:

A dummy argument is type, kind, and rank compatible, or TKR compatible, with another dummy argument if the first is type compatible with the second, the kind type parameters of the first have the same values as the corresponding kind type parameters of the second, and both have the same rank or either is assumed-rank.

File: lib/evaluate/characteristics.h [hunk 9]
```diff
@@ -44,26 +45,18 @@ extern template class Fortran::common::Indirection<
 
 namespace Fortran::evaluate::characteristics {
 
+template<typename T> using CopyableIndirection = common::CopyableIndirection<T>;
```
Comment (klausler) [inline lib/evaluate/characteristics.h]: `using common::CopyableIndirection;` might work.

Comment (tskeith) [reply lib/evaluate/characteristics.h]: You're right. I thought I had tried that but apparently not.

File: lib/evaluate/characteristics.cc (hunk 8 above)
Comment (tskeith) [reply lib/evaluate/characteristics.cc]: OK

File: lib/evaluate/type.h [hunk 10]
```diff
@@ -96,7 +96,11 @@ class DynamicType {
   }
   explicit constexpr DynamicType(
       const semantics::DerivedTypeSpec &dt, bool poly = false)
-    : category_{TypeCategory::Derived}, derived_{&dt} {}
+    : category_{TypeCategory::Derived}, derived_{&dt} {
+    if (poly) {
+      kind_ = ClassKind;
+    }
```
Comment (tskeith) [inline lib/evaluate/type.h]: @klausler, FYI, I had to tweak your change here for polymorphic types to work.

Comment (klausler) [reply lib/evaluate/type.h]: Thanks.

**Pull request merged**